### PR TITLE
fix(mempool): Propagate `CheckTx` errors to caller

### PR DESCRIPTION
**Pull request by hvanz**
Solves #1516

This PR changes `ReqRes`'s callback on the ABCI client to return an error. This error is then returned by the functions `handleCheckTxResponse` and `handleRecheckTxResponse`.  In case the tx is being added via an RPC endpoint, the error is propagated from `handleCheckTxResponse` up to the caller.

---

#### PR checklist

- [x] Tests written/updated
- [x] Changelog entry added in `.changelog` (we use [unclog](https://github.com/informalsystems/unclog) to manage our changelog)
- [ ] Updated relevant documentation (`docs/` or `spec/`) and code comments

### CODE REVIEW

File: abci/client/client.go [hunk 1]
```diff
@@ -115,18 +116,25 @@ func (r *ReqRes) InvokeCallback() {
...
+
 // GetCallback returns the configured callback of the ReqRes object which may be
 // nil. Note, it is not safe to concurrently call this in cases where it is
 // marked done and SetCallback is called before calling GetCallback as that
 // will invoke the callback twice and create a potential race condition.
 //
 // ref: https://github.com/tendermint/tendermint/issues/5439
-func (r *ReqRes) GetCallback() func(*types.Response) {
+//
+// Deprecated: not used.
```
Comment (melekes) [inline abci/client/client.go:136]: Why not remove it?

**Review by melekes — APPROVED** [review]
👍

File: abci/client/client.go (hunk 1 above)
Comment (hvanz) [reply abci/client/client.go:136]: Wouldn't that be API breaking? If yes, I'd remove it in the next major release.

Comment (melekes) [reply abci/client/client.go:136]: It is, but it's okay to break API in `v1` and `main`

Comment (hvanz) [reply abci/client/client.go]: Ok, I removed it.

Comment (hvanz) [reply abci/client/client.go]: Actually, since this is the only breaking part in this PR, I restored the method and created another PR to remove it #4083.

**Pull request merged**